llvm_objects.py

```diff
@@ -0,0 +1,238 @@
+            # and the two basic blocks to jump to if it is true or false
+            if self.is_conditional:
+               self.condition = LLVMGetCondition(l_instr)
+               self.l_bb_true = LLVMValueAsBasicBlock(self.l_operands[2])
+               self.l_bb_false = LLVMValueAsBasicBlock(self.l_operands[1])
+            # otherwise, store the basic block to jump to unconditionally
+            else:
+                self.l_bb_uncond = LLVMValueAsBasicBlock(self.l_operands[0])
+        elif self.opcode == LLVMICmp:
+            self.icmp_predicate = LLVMGetICmpPredicate(l_instr)
+        elif self.opcode == LLVMPHI:
+            self.count_incoming = LLVMCountIncoming(l_instr)
+            self.incoming_block = []
+            self.incoming_value = []
+            for i in range(self.count_incoming):
+                self.incoming_block.append(LLVMGetIncomingBlock(l_instr, i))
+                self.incoming_value.append(LLVMGetIncomingValue(l_instr, i))
+        elif self.opcode == LLVMCall:
+            if not LLVMIsDeclaration(self.l_operands[-1]):
+                self.func_param_count = LLVMCountParams(self.l_operands[-1])
+                self.l_func_params = []
+                for index in range(self.func_param_count):
+                    self.l_func_params.append(LLVMGetParam(self.l_operands[-1], index))
+            else:
+                # assume the instruction is either printf or puts (this is checked by the
+                # the interpreter)
+                self.l_string_format_ref = LLVMGetOperand(self.l_operands[0], 0)
+                self.func_name = rffi.charp2str(LLVMGetValueName(self.l_operands[-1]))
+
+    def get_operands(self, l_instr, w_module):
+        ''' Returns the operands of this instruction and stores the
+            constant ones in the given W_Module. '''
+
+        args = []
+        for i in range(LLVMGetNumOperands(l_instr)):
+            arg = LLVMGetOperand(l_instr, i)
+            args.append(arg)
+            if LLVMIsConstant(arg):
+                addr = rffi.cast(rffi.INT, arg)
+                var_type = LLVMGetTypeKind(LLVMTypeOf(arg))
+                if var_type == LLVMIntegerTypeKind:
+                    if LLVMIsUndef(arg):
+                        # XXX undefined vars -- currently set to 0
+                        w_module.constants.set_variable(addr, Integer(0))
+                    else:
+                        w_module.constants.set_variable(addr, Integer(LLVMConstIntGetSExtValue(arg)))
+                elif var_type == LLVMDoubleTypeKind or var_type == LLVMFloatTypeKind:
+                    if LLVMIsUndef(arg):
+                        w_module.constants.set_variable(addr, Float(0.0))
+                    else:
+                        with lltype.scoped_alloc(rffi.SIGNEDP.TO, 1) as signed_ptr:
+                            fl_var = Float(LLVMConstRealGetDouble(arg, signed_ptr))
+                            w_module.constants.set_variable(addr, fl_var)
+        return args
```

Comment:
in all the classes in this file, it would be good to separate the "construction-time methods" that use the LLVM API from the externally usable methods that the interpreter uses. Maybe move the constructions methods to after the ``__init__`` and separate the public ones with a comment? Or give the construction methods all an ``_`` prefix, since they are internal?